functions/supplier/s3-supplier/src/main/java/org/springframework/cloud/fn/supplier/s3/AwsS3SupplierConfiguration.java

```diff
@@ -37,80 +44,198 @@
+			return synchronizer;
+		}
+
+		@Bean
+		public MessageSource<File> s3MessageSource(S3InboundFileSynchronizer s3InboundFileSynchronizer) {
+			S3InboundFileSynchronizingMessageSource s3MessageSource = new S3InboundFileSynchronizingMessageSource(
+					s3InboundFileSynchronizer);
+			s3MessageSource.setLocalDirectory(this.awsS3SupplierProperties.getLocalDir());
+			s3MessageSource.setAutoCreateLocalDirectory(this.awsS3SupplierProperties.isAutoCreateLocalDir());
+			return s3MessageSource;
+		}
 	}
 
-	@Bean
-	public Supplier<Flux<Message<?>>> s3Supplier() {
-		return () -> Flux.from(s3SupplierFlow());
+	@Configuration
+	@ConditionalOnExpression("environment['s3.supplier.list-only'] == 'true'")
+	static class ListOnlyConfiguration extends AwsS3SupplierConfiguration {
+		ListOnlyConfiguration(AwsS3SupplierProperties awsS3SupplierProperties,
+				FileConsumerProperties fileConsumerProperties,
+				AmazonS3 amazonS3,
+				ResourceIdResolver resourceIdResolver, ConcurrentMetadataStore metadataStore) {
+			super(awsS3SupplierProperties, fileConsumerProperties, amazonS3, resourceIdResolver, metadataStore);
+		}
+
+		@Bean
+		public Publisher<Message<Object>> s3SupplierFlow(MessageSource<?> s3MessageSource,
+				GenericSelector<S3ObjectSummary> listOnlyFilter) {
+			return IntegrationFlows
+					.from(IntegrationReactiveUtils.messageSourceToFlux(s3MessageSource))
+					.filter(listOnlyFilter)
+					.toReactivePublisher();
+		}
+
+		@Bean
+		PollableChannel listingChannel() {
+			return new QueueChannel();
+		}
+
+		@Bean
+		GenericSelector<S3ObjectSummary> listOnlyFilter() {
+			Predicate<S3ObjectSummary> predicate = s -> true;
+			if (StringUtils.hasText(this.awsS3SupplierProperties.getFilenamePattern())) {
+				Pattern pattern = Pattern.compile(this.awsS3SupplierProperties.getFilenamePattern());
+				predicate = (S3ObjectSummary summary) -> pattern.matcher(summary.getKey()).matches();
+			}
+			else if (this.awsS3SupplierProperties.getFilenameRegex() != null) {
+				predicate = (S3ObjectSummary summary) -> this.awsS3SupplierProperties.getFilenameRegex()
+						.matcher(summary.getKey()).matches();
+			}
+			predicate = predicate.and((S3ObjectSummary summary) -> {
+				final String key = METADATA_STORE_PREFIX + summary.getBucketName() + "-" + summary.getKey();
+				final String lastModified = String.valueOf(summary.getLastModified().getTime());
+				final String storedLastModified = this.metadataStore.get(key);
+				boolean result = !lastModified.equals(storedLastModified);
+				if (result) {
+					metadataStore.put(key, lastModified);
+				}
+				return result;
+			});
+
+			GenericSelector<S3ObjectSummary> selector = predicate::test;
+
+			return selector;
+		}
+
+		@Bean
+		public MessageSource<?> s3MessageSource(PollableChannel listingChannel,
+				S3ListingMessageProducer s3ListingMessageProducer) {
+			return () -> {
+				s3ListingMessageProducer.getObjectMetadata();
+				return (Message<Object>) listingChannel.receive();
```

Comment:
I wonder if you are OK to revise the solution in favor of `ReactiveMessageSourceProducer`.

I mean you leave the solution for `list` only with that `S3ListingMessageProducer`, but for other `MessageSource`-based, use this `ReactiveMessageSourceProducer` instead of that `IntegrationReactiveUtils.messageSourceToFlux()`.

Well, in fact I see you really did an extra layer converting `MessageProducerSupport` to `MessageSource` and than to the `Flux` in the flow.
You definitely need just an `IntegrationFlows.from(MessageProducerSupport)` for this `list` variant.

Reply:
I get no poller configured when I do that